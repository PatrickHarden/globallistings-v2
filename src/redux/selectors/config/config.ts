import { createSelector } from 'reselect';
import { configSelector, configMetaSelector } from '../state';

export const configStatusSelector = createSelector(
  configMetaSelector,
  config => config.status
);

export const getFilters = createSelector(
  configSelector,
  config => config.filters
);

export const siteTypeSelector = createSelector(
  configSelector,
  config => config.siteType
);
export const searchModeSelector = createSelector(
  configSelector,
  config => config.searchMode
);

// SearchConfig
export const searchConfigSelector = createSelector(
  configSelector,
  config => config.searchConfig
);

export const searchBiasSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchBias
);

export const searchRadiusRestrictionSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.biasRadius
);

export const searchCountryRestrictionSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.restrictToCountry
);

export const polygonSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.polygons
);

export const searchTypesRestrictionSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchPlaceTypes
);

export const assumeZipCodeSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.allowAssumeZipCode
);

export const searchPathSortAlphabeticalSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchPathSortAlphabetical
);

export const searchPathSelectorSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchPathSelector
);

export const propertyIdPathSelectorSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.propertyIdSearchPathSelector
);

export const hideBuyButtonSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.hideSearchToBuy
);

export const hideLetButtonSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.hideSearchToLet
);

export const getSearchHeader = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchHeader
);
export const searchResultsPageSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchResultsPage
);

export const searchSuggestDebounceSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchSuggestDebounce
);

export const defaultSearchLocationNameSelector = createSelector(
  configSelector,
  config => config.searchLocationName
);

export const paramsSelector = createSelector(
  configSelector,
  config => config.params
);

export const restrictionsSelector = createSelector(
  configSelector,
  config => ({
    country: config.searchConfig.restrictToCountry,
  })
);

export const otherConfigsSelector = createSelector(
  configSelector,
  config => (config.otherConfigs ? config.otherConfigs : [])
);

export const currentConfigSelector = createSelector(
  configSelector,
  config => config.currentConfig
);
