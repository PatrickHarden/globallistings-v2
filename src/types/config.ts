import { Dictionary } from './common/dictionary';

interface BaseConfig {
  readonly api: string;
  readonly bitlyToken: string;
  readonly breakpoints: Breakpoints;
  readonly cdnUrl: string;
  readonly filters: Filter[];
  readonly language: string;
  readonly largePlaceholderImageUrl: string;
  readonly mapMarkerIconUrl: string;
  readonly mapMarkerIconUrlInactive: string;
  readonly mediumPlaceholderImageUrl: string;
  readonly params: Params;
  readonly pdf: Pdf;
  readonly pdpContentDrivenBlock?: any;
  readonly pdpContentDrivenBlockLettings?: any;
  readonly placesAddressLookupStringFormat: string;
  readonly propertyContactApiUrl: string;
  readonly recaptchaKey: string;
  readonly searchConfig: SearchConfig;
  readonly searchLocationName?: string;
  readonly searchMode: string;
  readonly siteId: string;
  readonly siteType: string;
  readonly smallPlaceholderImageUrl: string;
  readonly tenantFeesLinkUrl?: string;
  readonly title: string;
  readonly urlPropertyAddressFormat: string;
  readonly currentConfig?: string;
  readonly limitListMapResults?: string;
  readonly otherConfigs?: OtherConfig[];
}

/**
 * These types should only be used by selectors and not by components directly
 * selectors need to output their own types
 */

export interface Config extends BaseConfig {
  readonly i18n: object;
}

export interface AppConfig extends BaseConfig {
  readonly i18n: Dictionary<string>;
}

export interface FilterOption {
  readonly value: string;
  readonly label: string;
  readonly default: boolean;
  readonly type?: string;
}

export enum FilterPlacement {
  group = 'lm_groupFilters',
  location = 'lm_locationFilter',
  primary = 'lm_primaryFilter',
  searchPrimary = 'primary',
  searchSecondary = 'secondary',
  secondary = 'lm_secondaryFilter',
  sort = 'lm_sortFilter',
  tertiary = 'lm_tertiaryFilters',
}

export enum FilterType {
  select = 'select',
  range = 'range',
}

export enum FilterModifier {
  query = 'query',
  url = 'url',
}

export type FilterCondition = { [K in keyof Params]?: string };

export interface BaseFilter {
  readonly placement: FilterPlacement;
  readonly modifier: FilterModifier;
  readonly name: string;
  readonly label: string;
  readonly conditional?: FilterCondition;
}

export interface SelectFilter extends BaseFilter {
  readonly options: FilterOption[];
  readonly type: FilterType.select;
}

export interface RangeFilter extends BaseFilter {
  readonly includePOA?: boolean;
  readonly maxValues: FilterOption[];
  readonly minValues: FilterOption[];
  readonly type: FilterType.range;
}

export type Filter = SelectFilter | RangeFilter;

interface Breakpoints {
  readonly xsmall: string;
  readonly small: string;
  readonly medium: string;
  readonly large: string;
}

interface Pdf {
  readonly staticMapApi: string;
  readonly pdfDownloadApi: string;
  readonly renderMiddleware: string;
}

export interface Params {
  readonly 'Common.Aspects': string;
  readonly 'Common.UsageType': string;
  readonly CurrencyCode: string;
  readonly 'Dynamic.TotalArea': string;
  readonly Interval: string;
  readonly PageSize: string;
  readonly RadiusType: string;
  readonly Site: string;
  readonly PropertyDefaultEnumerationSuffixes?: string[];
  readonly PropertyOrigins?: string[];
  readonly PropertyQualifier?: string;
  readonly Sort: string;
  readonly Unit: string;
}

export interface Polygon {
  coords: string,
  value: string,
  label: string,
  default?: boolean,
  name?: string
}

export interface SearchConfig {
  readonly searchResultsPage: string; // Can't be an empty string
  readonly searchHeaderLinkUrl?: string; // Can't be an empty string
  readonly searchBias?: SearchBias; // If it's defined biasRadius is compulsory
  readonly biasRadius?: number;
  readonly restrictToCountry?: string;
  readonly searchPlaceTypes?: string[]; // Can't be an empty array
  readonly allowAssumeZipCode?: boolean; // if this is present and true, then at 4-5 digits we add (regions) to types to eliminate IP location bias
  readonly searchPathSortAlphabetical?: boolean; // sort the search path selector alphabetically (i.e, property types)
  readonly searchHeader: SearchHeader;
  readonly hideSearchToBuy: boolean;
  readonly hideSearchToLet: boolean;
  readonly searchPathSelector?: SearchPath[]; // Can't be an empty array
  readonly propertyIdSearchPathSelector?: PropertyIdSearchPath[]; // Can't be an empty array
  readonly searchSuggestDebounce?: number;
  readonly polygons?: Polygon[];
}

export interface SearchPath {
  readonly value: string;
  readonly label: string;
  readonly default: boolean;
}

export interface PropertyIdSearchPath {
  readonly value: string;
  readonly label: string;
  readonly display?: string;
}

export interface SearchHeader {
  readonly searchHeaderTitleText?: string;
  readonly searchHeaderLinkText?: string;
  readonly searchHeaderLinkUrl?: string;
}

export interface SearchBias {
  readonly lat: string;
  readonly lon: string;
}

export interface LocationTypeDefinition {
  readonly name: string;
  readonly definitions: string[];
  readonly radius: number;
}

export interface OtherConfig {
  readonly name: string;
  readonly endpoint: string;
}

export enum TransactionTypeValue {
  isSale = 'isSale',
  isLetting = 'isLetting',
}

export enum DistanceUnit {
  miles = 'miles',
  km = 'km',
}
