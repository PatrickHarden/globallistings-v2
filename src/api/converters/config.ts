import { mapObjIndexed } from 'ramda';
import { Config } from '../../types/config';
import flattenObject from '../../utils/flattenObj';

const replaceIntlChars = mapObjIndexed((_, key, obj) =>
  obj[key].replace(/%\(/g, '{').replace(/\)s/g, '}')
);

const convertToIntl = (messages: object) => replaceIntlChars(messages);

// TODO this should leave outside API folder as this is not an API layer anymore and just a transformation layer
export const convert = (config: Config): Config => ({
  ...config,
  i18n: convertToIntl(flattenObject(config.i18n)),
});
