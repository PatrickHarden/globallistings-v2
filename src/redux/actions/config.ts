import axios from 'axios';
import {
  createRequestPending,
  createRequestFailure,
  createRequestSuccess,
} from '../../types/common/action-creators-singular';
import { Config } from '../../types/config';
import { State } from '../../types/state';
import { configEndpointForCurrentUsageTypeSelector } from '../selectors/plp';

export const createFetchConfigRequest = createRequestPending('FETCH_CONFIG_REQUEST');
export const createFetchConfigFailure = createRequestFailure('FETCH_CONFIG_FAILURE');
export const createFetchConfigSuccess = createRequestSuccess<Config, 'FETCH_CONFIG_SUCCESS'>(
  'FETCH_CONFIG_SUCCESS'
);

// This will be used in the future to switch between multiple configs
export const fetchConfig = () => (dispatch: Function, getState: () => State) => {
  const state = getState();
  const configEndpoint = configEndpointForCurrentUsageTypeSelector(state);

  dispatch(createFetchConfigRequest());
  return axios
    .get(configEndpoint)
    .then(res => {
      dispatch(createFetchConfigSuccess(res.data));
    })
    .catch(ex => {
      dispatch(createFetchConfigFailure(ex));
    });
};
