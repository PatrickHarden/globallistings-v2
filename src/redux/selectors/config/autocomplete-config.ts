import { createSelector } from 'reselect';
import { Dictionary } from '../../../types/common/dictionary';
import { PropertyIdSearchPath } from '../../../types/config';

import {
  paramsSelector,
  searchPathSelectorSelector,
  propertyIdPathSelectorSelector,
  searchBiasSelector,
  searchRadiusRestrictionSelector,
  searchCountryRestrictionSelector,
  searchTypesRestrictionSelector,
  assumeZipCodeSelector,
} from './config';
import { messagesSelector } from '../state';

export interface PropertySearchConfig {
  readonly Site: string;
  readonly PropertyOrigins?: string;
  readonly SearchPaths?: PropertyIdSearchPath[];
  readonly Messages?: Dictionary<string>;
}

export const propertySearchConfig = createSelector(
  paramsSelector,
  propertyIdPathSelectorSelector,
  (params, searchPaths): PropertySearchConfig => {
    return {
      Site: params.Site,
      PropertyOrigins: params.PropertyOrigins![0] || '',
      SearchPaths: searchPaths,
    };
  }
);

export interface AutocompleteConfig {
  readonly bias?: {
    readonly lat: number;
    readonly lng: number;
    readonly radius: number;
  };
  readonly country?: string;
  readonly types?: Array<string>;
  readonly allowAssumeZipCode?: boolean;
}

export const placeAutoCompleteConfig = createSelector(
  searchBiasSelector,
  searchRadiusRestrictionSelector,
  searchCountryRestrictionSelector,
  searchTypesRestrictionSelector,
  assumeZipCodeSelector,
  (bias, radius, country, types, allowAssumeZipCode): AutocompleteConfig => {
    // JS XOR logic
    if (bias ? !radius : radius) {
      throw new Error('Both bias and radius must be present in the config or none at all');
    }

    // TS doesn't correctly understand XOR so we have to check for both bias and radius
    if (!bias || !radius) {
      if (allowAssumeZipCode) {
        return { country, types, allowAssumeZipCode };
      } else {
        return { country, types };
      }
    }

    return {
      bias: {
        lat: Number.parseFloat(bias.lat),
        lng: Number.parseFloat(bias.lon),
        radius,
      },
      country,
      types,
      allowAssumeZipCode,
    };
  }
);
