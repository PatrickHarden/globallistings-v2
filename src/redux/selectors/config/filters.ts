import { filtersSelector } from '../state';
import { createSelector } from 'reselect';

import { querySelector, locationParamsSelector } from '../location';
import { FilterOption, FilterType, BaseFilter } from '../../../types/config';
import { Dictionary } from '../../../types/common/dictionary';

export interface SelectInfo {
  readonly items: FilterOption[];
  readonly selectedItem?: FilterOption;
}

export interface SelectFilterInfo extends BaseFilter, SelectInfo {
  readonly type: FilterType.select;
}

export interface RangeFilterInfo extends BaseFilter {
  readonly type: FilterType.range;
  readonly min: SelectInfo;
  readonly max: SelectInfo;
}

export type FilterInfo = SelectFilterInfo | RangeFilterInfo;

const getValidRangeOptions = (
  options: FilterOption[],
  selectedOption: FilterOption | undefined,
  comparator: (a: number, b: number) => boolean
) => {
  if (!selectedOption) {
    return options;
  }

  return options.filter(({ value }) => {
    // Allow default values
    if (value === '') {
      return true;
    }
    const a = parseInt(value, undefined);
    const b = parseInt(selectedOption.value, undefined);
    if (isNaN(a) || isNaN(b)) {
      console.warn(
        `${value} cannot be parsed to a valid integer therefore disabling of invalid range filter values will not work.`
      );
      return true;
    }
    return comparator(a, b);
  });
};

const filterOptionToQueryValue = (filterName: string, query: any) => (f: FilterOption) =>
  f.value === query[`${filterName}_${f.type}`];

const findQueryValue = (filterName: string, query: any, key: string) => {
  const keyObj = query[key];
  return keyObj !== undefined ? keyObj[filterName] : query[filterName];
};

export const filtersInfoSelector = createSelector(
  filtersSelector,
  locationParamsSelector,
  (filters, query): FilterInfo[] =>
    filters
      .filter(f => [FilterType.range, FilterType.select].indexOf(f.type) > -1)
      .map(f => {
        const queryValue = findQueryValue(f.name, query, 'params');
        if (Array.isArray(queryValue)) {
          throw new Error(`query param ${queryValue} must not be an array`);
        }

        if (f.type === FilterType.select) {
          const value = f.options.find(o => o.value === queryValue);
          return {
            ...f,
            items: f.options,
            selectedItem: value,
          };
        }
        if (f.type === FilterType.range) {
          const minValue = f.minValues.find(filterOptionToQueryValue(f.name, query));
          const maxValue = f.maxValues.find(filterOptionToQueryValue(f.name, query));
          const minValues = getValidRangeOptions(f.minValues, maxValue, (a, b) => a <= b);
          const maxValues = getValidRangeOptions(f.maxValues, minValue, (a, b) => a >= b);

          return {
            range: FilterType.range,
            type: f.type,
            placement: f.placement,
            label: f.label,
            name: f.name,
            modifier: f.modifier,
            min: {
              items: minValues,
              selectedItem: minValue,
            },
            max: {
              items: maxValues,
              selectedItem: maxValue,
            },
          };
        }

        throw new Error('Unsupported filter type');
      })
);

export const defaultFiltersParamsSelectors = createSelector(
  filtersSelector,
  (filters): Dictionary<string> =>
    filters
      .map(f => {
        if (f.type === FilterType.select) {
          const defaultOption = f.options.find(o => o.default);
          if (!defaultOption) {
            return {};
          }
          return { [f.name]: defaultOption.value };
        }
        // TODO add support for range filters
        return {};
      })
      .reduce((prev, curr) => ({ ...prev, ...curr }), {})
);
