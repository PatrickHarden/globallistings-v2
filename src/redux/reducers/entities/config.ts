import { convert } from '../../../api/converters/config';
import { createFetchConfigSuccess } from '../../actions/config';
import { ActionType } from 'typesafe-actions';
import { DeepPartial } from 'redux';
import { Config } from '../../../types/config';

export type Actions = ActionType<typeof createFetchConfigSuccess>;

export default (state: DeepPartial<Config> = {}, action: Actions): Config | {} => {
  switch (action.type) {
    case 'FETCH_CONFIG_SUCCESS':
      return convert(action.payload);
    default:
      return state;
  }
};
