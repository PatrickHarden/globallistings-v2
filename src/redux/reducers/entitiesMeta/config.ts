import * as actions from '../../actions/config';
import { ActionType } from 'typesafe-actions';
import { handleError, handleSuccess, handlePending } from './shared/meta-reducers';

export type Actions = ActionType<typeof actions>;

export default (state = {}, action: Actions) => {
  switch (action.type) {
    case 'FETCH_CONFIG_REQUEST':
      return handlePending();
    case 'FETCH_CONFIG_SUCCESS':
      return handleSuccess(action);
    case 'FETCH_CONFIG_FAILURE':
      return handleError(action);
    default:
      return state;
  }
};
