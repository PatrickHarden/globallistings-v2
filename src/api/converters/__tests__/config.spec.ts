import { convert } from '../config';
import { Config } from '../../../types/config';
import { DeepPartial } from '../../../types/common/deep-partial';

describe(convert, () => {
  it('should correctly convert i18n', () => {
    const config: DeepPartial<Config> = {
      i18n: {
        one: 'uno',
        two: '%(two)s - dos',
        three: '%(three)s - tres %(three)s, (%(three)s)',
        nested: {
          four: '%(cuatro)s',
          moreNested: {
            five: '%(cinco)s',
          },
        },
      },
    };

    expect(convert(config as Config)).toMatchSnapshot();
  });

  it('should correctly overwrite defaults', () => {
    const config: DeepPartial<Config> = {
      i18n: {},
      searchConfig: {
        hideSearchToBuy: true,
        searchHeader: {
          searchHeaderTitleText: 'zap',
        },
        searchBias: {
          lat: '0',
          lon: '0',
        },
      },
    };

    expect(convert(config as Config)).toMatchSnapshot();
  });
});
