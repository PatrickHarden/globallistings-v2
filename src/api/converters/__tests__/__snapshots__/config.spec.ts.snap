// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert should correctly convert i18n 1`] = `
Object {
  "i18n": Object {
    "five": "{cinco}",
    "four": "{cuatro}",
    "one": "uno",
    "three": "{three} - tres {three}, ({three})",
    "two": "{two} - dos",
  },
}
`;

exports[`convert should correctly overwrite defaults 1`] = `
Object {
  "i18n": Object {},
  "searchConfig": Object {
    "hideSearchToBuy": true,
    "searchBias": Object {
      "lat": "0",
      "lon": "0",
    },
    "searchHeader": Object {
      "searchHeaderTitleText": "zap",
    },
  },
}
`;
